### ISSUE TITLE
Is the description for --process-gcno wrong?

### ISSUE THREAD

**Issue by varunsh-xilinx** [RPGillespie6/fastcov#77]
The description reads "Process both gcno and gcda coverage files. This option is useful for capturing untested files in the coverage report." I took it to mean that it would do some additional processing on .gcno files for extra information on top of the default .gcda processing. However, it looks like the [findCoverageFiles()](https://github.com/RPGillespie6/fastcov/blob/master/fastcov.py#L246) function considers only one of *.gcda or *.gcno depending on the value of the `--process-gcno` flag. 

I also ran into an error when using this flag in my Cmake setup. I passed in the `--zerocounters` flag thinking that it would remove the .gcda files but when this is used with `--process-gcno`, it actually removes all the .gcno files instead. The logging output is also misleading here because it always reads that "Removed X .gcda files" independent of what was actually removed. The issue was that Cmake wouldn't regenerate the .gcno files unless I do a clean build so I get an error when fastcov can't find any .gcno files.

```
[1.774s] [info]: Found 52 coverage files (gcno)
[0.146s] [info]: Removed 52 .gcda files

...

[0.886s] [info]: Found 0 coverage files (gcno)
[0.022s] [info]: Spawned 0 gcov processes, each processing at most 5 coverage files
Traceback (most recent call last):
  File "/usr/local/bin/fastcov", line 994, in <module>
    main()
  File "/usr/local/bin/fastcov", line 968, in main
    fastcov_json = getGcovCoverage(args)
  File "/usr/local/bin/fastcov", line 824, in getGcovCoverage
    fastcov_json = processGcdas(args, coverage_files, gcov_filter_options)
  File "/usr/local/bin/fastcov", line 319, in processGcdas
    base_fastcov = fastcov_jsons.pop()
IndexError: pop from empty list
```

**Comment by RPGillespie6** [comment]
Yes, you are right, I think this is a bug (sorry, I don't use this feature much myself or I probably would have caught it earlier). I think the behavior of fastcov should be changed as follows:
- if `--process-gcno` is present, process _both_ gcno _and_ gcda
- if `--zerocounters` is passed, only delete gcda, not gcno
- If no coverage files are found, exit early without doing anything

That should account for all the issues you brought up, I believe.

**Comment by RPGillespie6** [comment]
As a temporary workaround, you could do 2 runs:
- one with `--process-gcno` and one without
- combine the two reports with `fastcov.py -C`

**Comment by varunsh-xilinx** [comment]
The two-pass approach works well for me.

The actions you listed sound good to me. Another option is that you can disallow using both `--zerocounters` and `--process-gcno` at the same time (either in documentation or at the argparse level) and require two passes to do this. I think that's a fine approach as well.

**Comment by johan-boule** [comment]
Are you sure `--process-gcno` doesn't delete the `.gcno` files regardless of the `zerocounters` options ?
I don't see a reason to delete the `.gcno` files as these are compiler output, not test execution byproducts.

**Comment by johan-boule** [comment]
Indeed, `.gcno` files should never be deleted, as these are precious products of the compiler, not of the test execution.
Isn't `--zerocounters` just a fancy `find "$build" -name '*.gcda' -delete` ?
A related question, does fastcov avoid creating .gcov files by piping in the output of gcov ?

**Comment by RPGillespie6** [comment]
> I don't understand the two-pass approach with merging

The two pass approach is a workaround due to the bugs identified in this issue.

> Isn't the run with all gcno files a superset of the run

It's _supposed_ to be, but this issue is open because currently passing `--process-gcno` makes you process _only_ gcno, not both gcno and gcda:

See: https://github.com/RPGillespie6/fastcov/blob/master/fastcov.py#L249
```python
        coverage_type = "gcno" if use_gcno else "gcda"
        coverage_files = glob.glob(os.path.join(os.path.abspath(cwd), "**/*." + coverage_type), recursive=True)
```

> Isn't --zerocounters just a fancy find "$build" -name '*.gcda' -delete that's been brought from lcov ?

Yes, but this issue is open because if you pass both `--zerocounters` and `--process-gcno` fastcov deletes all the gcno files because of the referenced lines above.

> A related question, does fastcov avoid creating .gcov files by piping in the output of gcov ?

Yes, fastcov avoids creating .gcov files using gcov's `-t/--stdout` flag which pipes coverage reports directly to stdout, which fastcov parses.

Anyway, I'll fix this issue someday, I've been super busy lately and haven't found the time. PRs are welcome if anyone wants to do it.

**Comment by johan-boule** [comment]
I gave it a shot. Please see the pull request above.

**Comment by RPGillespie6** [comment]
@varunsh-xilinx The `--process-gcno` flag behaves as expected. You did uncover one issue though - as you and @johan-boule noted, gcno files should never be deleted by fastcov. I've pushed a fix for that in #80. 

The way gcov works is that regardless of the file extensions of the files you pass to gcov, it will strip off the extension and then look for both `gcno` and `gcda` variants. So fundamentally, what passing `--process-gcno` does is force inclusion of sources files that were not touched by tests in the final report. If you do not pass `--process-gcno`, only files touched by tests will be included in the final report.

**Comment by johan-boule** [comment]
Thank you very much for having dug deeper in the analysis and for the final fix.
